### PR TITLE
Support passing a json or yaml string as file param to validate definition

### PR DESCRIPTION
**Pull request by joejstuart**
- Support json or yaml string as `--file` param
- Remove extra afero mock

### CODE REVIEW

File: internal/definition/validate.go [hunk 1]
```diff
@@ -93,3 +98,31 @@ func detectFiles(ctx context.Context, fpath string) ([]string, error) {
 
 	return defFiles, nil
 }
+
+// detect if the string is json
+func detectJson(msg string) bool {
```
Comment (zregvart) [inline internal/definition/validate.go:103]: We have similar logic in a couple of places, perhaps a good candidate to place in the `internal/utils`?

```suggest
func IsJson(text string) bool
```

File: internal/definition/validate.go [hunk 2]
```diff
@@ -93,3 +98,31 @@ func detectFiles(ctx context.Context, fpath string) ([]string, error) {
...
+		return nil, err
+	}
+
+	filePath := filepath.Join(tmpDir, fileName)
+	err = afero.WriteFile(fs, filePath, []byte(msg), 0644)
+	if err != nil {
+		return nil, err
+	}
+	return []string{filePath}, nil
+}
```
Comment (zregvart) [inline internal/definition/validate.go:128]: Perhaps something like:
```suggestion
func writeTempFile(ctx context.Context, data string) (string, error) {
	fs := utils.FS(ctx)
	file, err := fs.TempFile("", "*")
	if err != nil {
		return "", err
	}
	path := file.Name()
	if _, err := file.WriteString(data); err != nil {
		_ := fs.Remove(path)
		return "", err
	}
	return path, nil
}
```

File: internal/utils/helpers.go [hunk 3]
```diff
@@ -98,3 +100,16 @@ func FS(ctx context.Context) afero.Fs {
 func WithFS(ctx context.Context, fs afero.Fs) context.Context {
 	return context.WithValue(ctx, fsKey, fs)
 }
+
+func GenerateRandomString(length int) (string, error) {
```
Comment (zregvart) [inline internal/utils/helpers.go:104]: Guess we might not need this?

**Comment by zregvart** [comment]
Would we also want to support YAML?

**Comment by lcarva** [comment]
> Would we also want to support YAML?

I think at this point, we should do this by default. We'd need a reason **not** to do it.

File: internal/definition/validate.go (hunk 1 above)
Comment (joejstuart) [reply internal/definition/validate.go]: Good point. I'm sure that can be accommodated.

File: internal/definition/validate.go [hunk 4]
```diff
@@ -55,44 +55,73 @@ func ValidateDefinition(ctx context.Context, fpath string, sources []source.Poli
 }
 
 // detect if a file or directory was passed. if a directory, gather all files in it
-func detectFiles(ctx context.Context, fpath string) ([]string, error) {
+// the order is file lookup, json lookup then yaml
+func detectInput(ctx context.Context, fpath string) ([]string, error) {
+	log.Error("trying to detect input as JSON")
```
Comment (zregvart) [inline internal/definition/validate.go:60]: Not sure about logging at this level, perhaps the only error we should print is when we can't detect/load from a path?

**Review by zregvart — APPROVED** [review]
Straightforward enough, not sure about the error logging level

File: internal/definition/validate.go (hunk 4 above)
Comment (joejstuart) [reply internal/definition/validate.go]: I wasn't in love with it either. Was just trying to find a way to communicate to the user what was going on. I moved the logging to after the validation attempt and to debug. wdyt?

```golang
if utils.IsJson(fpath) {
    log.Debug("valid JSON found for definition file")
    return definitionFromString(ctx, fpath)
}
log.Debug("unable to detect input as JSON")
```

**Review by lcarva — APPROVED** [review]

File: internal/definition/validate.go [hunk 5]
```diff
@@ -55,41 +55,81 @@ func ValidateDefinition(ctx context.Context, fpath string, sources []source.Poli
...
+	log.Debug("unable to detect input as YAML")
+
+	files, err := fileLookup(ctx, fpath)
+	if len(files) > 0 {
+		log.Debug("valid file path found for definition file")
+		return files, nil
+	}
+	// just log this error. it could be an actual os error or "file not found".
+	// either way, move on
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (joejstuart) [reply internal/definition/validate.go:75]: Good points. I think this is the best I can do with it. https://github.com/joejstuart/ec-cli/commit/d9fb3330ad2ec583b99f0cc04fd85da34aa15852

**Pull request merged**